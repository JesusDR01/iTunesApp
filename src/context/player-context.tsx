import { Podcast, Podcasts } from '@modules/podcasts/domain/Podcast';
import {
	Dispatch,
	MutableRefObject,
	createContext,
	useContext,
	useReducer,
	useRef,
} from 'react';

export type PodcastList = Podcasts;


export type State = {
	randomMode: boolean;
	currentPodcast: Podcast | null;
	podcastList?: PodcastList;
	isPlaying: boolean;
	volume: number;
	time: number;
	loopMode: 'all' | 'one' | 'none';
};

export type Action =
	| { type: 'play' }
	| { type: 'pause' }
	| { type: 'setPodcast'; podcast: Podcast }
	| { type: 'setPodcastList'; podcastList: PodcastList }
	| { type: 'loopMode'; loopMode: 'all' | 'one' | 'none' }
	| { type: 'randomMode'; randomMode: boolean }
	| { type: 'next' }
	| { type: 'end' }
	| { type: 'prev' }
	| { type: 'volume'; volume: number };

export type DispatchAction = React.Dispatch<Action>;

interface SearchPodcastContext {
	audioRef: MutableRefObject<HTMLAudioElement | null>;
	useControls: [State, Dispatch<Action>];
}

export const PlayerContext = createContext<SearchPodcastContext | null>(null);

export const PlayerContextProvider = ({
	children,
}: {
	children: React.ReactNode;
}) => {
	const audioRef = useRef<HTMLAudioElement | null>(null);
	const getInitialState = (): State => ({
		randomMode: false,
		currentPodcast: null,
		podcastList: undefined,
		isPlaying: false,
		volume: 0.3,
		time: 0,
		loopMode: 'none',
	});

	return (
		<PlayerContext.Provider
			value={{
				audioRef,
				useControls: useReducer((state: State, action: Action) => {
					switch (action.type) {
						case 'setPodcast': {
							const podcast = action.podcast;
							return {
								...state,
								currentPodcast: podcast,
							};
						}
						case 'setPodcastList': {
							const podcastList = action.podcastList;
							return {
								...state,
								podcastList: podcastList,
							};
						}
						case 'loopMode': {
							return { ...state, loopMode: action.loopMode };
						}
						case 'randomMode': {
							const randomMode = action.randomMode;
							return {
								...state,
								randomMode: randomMode,
							};
						}
						case 'prev': {
							const { podcastList, currentPodcast, loopMode } = state;
							if (loopMode === 'one' && audioRef.current) {
								audioRef.current.currentTime = 0;
								return {
									...state,
								};
							}
							if (podcastList && currentPodcast) {
								const currentPodcastIndex = podcastList.findIndex(
									podcast => podcast.episodeUrl === currentPodcast?.episodeUrl,
								);
								const prevPodcast = podcastList[currentPodcastIndex - 1];

								return {
									...state,
									currentPodcast: prevPodcast || currentPodcast,
								};
							}

							return {
								...state,
							};
						}
						case 'next': {
							const { podcastList, currentPodcast, randomMode, loopMode } =
								state;

							if (podcastList && currentPodcast) {
								const currentPodcastIndex = podcastList.findIndex(
									podcast => podcast.episodeUrl === currentPodcast?.episodeUrl,
								);

								const nextPodcast =
									podcastList[
										randomMode
											? Math.floor(Math.random() * podcastList.length)
											: currentPodcastIndex + 1
									];
								return {
									...state,
									currentPodcast: nextPodcast || currentPodcast,
									loopMode:
										loopMode === 'one' || loopMode === 'all'
											? ('all' as const)
											: ('none' as const),
								};
							}

							return {
								...state,
							};
						}
						case 'end': {
							const { podcastList, currentPodcast, randomMode, loopMode } =
								state;

								if (loopMode === 'one' && audioRef.current) {
									audioRef.current.currentTime = 0;
									audioRef.current.play();
									return {
										...state,
									};
								}

							if (podcastList && currentPodcast) {
								const currentPodcastIndex = podcastList.findIndex(
									podcast => podcast.episodeUrl === currentPodcast?.episodeUrl,
								);

								const nextPodcast =
									podcastList[
										randomMode
											? Math.floor(Math.random() * podcastList.length)
											: currentPodcastIndex + 1
									];
								return {
									...state,
									currentPodcast: nextPodcast || currentPodcast,
								};
							}

							return {
								...state,
							};
						}
						case 'play': {
							audioRef.current?.play();
							return { ...state, isPlaying: true };
						}
						case 'pause': {
							audioRef.current?.pause();
							return { ...state, isPlaying: false };
						}
						case 'volume': {
							const volume = action.volume;
							if (audioRef.current) {
								audioRef.current.volume = volume;
							}
							return {
								...state,
								volume: volume,
							};
						}
						default: {
							return state;
						}
					}
				}, getInitialState()),
			}}
		>
			{children}
		</PlayerContext.Provider>
	);
};

export const usePlayer = () => {
	const context = useContext(PlayerContext);

	if (context === null) {
		throw new Error('SearchPodcastContext error');
	}

	return context;
};
