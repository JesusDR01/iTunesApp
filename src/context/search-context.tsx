import {
	Dispatch,
	SetStateAction,
	createContext,
	useContext,
	useState,
} from 'react';

interface SearchPodcastContext {
	useSearch: [string, Dispatch<SetStateAction<string>>];
}

export const SearchPodcastContext = createContext<SearchPodcastContext | null>(
	null,
);

export const SearchPodcastsProvider = ({
	children,
}: {
	children: React.ReactNode;
}) => {
	const [search, setSearch] = useState('');

	return (
		<SearchPodcastContext.Provider
			value={{
				useSearch: [search, setSearch],
			}}
		>
			{children}
		</SearchPodcastContext.Provider>
	);
};

export const useSearch = () => {
	const context = useContext(SearchPodcastContext);

	if (context === null) {
		throw new Error('SearchPodcastContext error');
	}

	return context;
};
