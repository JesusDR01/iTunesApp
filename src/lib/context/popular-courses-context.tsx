import {
  Dispatch,
  SetStateAction,
  createContext,
  useContext,
  useState,
} from "react"

interface PopularCoursesContext {
  useDefaultValue: [string, Dispatch<SetStateAction<string>>]
}

export const PopularCoursesContext =
  createContext<PopularCoursesContext | null>(null)

export const PopularCoursesProvider = ({
  children,
}: {
  children: React.ReactNode
}) => {
  const [screen, setScreen] = useState("")
  return (
    <PopularCoursesContext.Provider
      value={{
        useDefaultValue: [screen, setScreen],
      }}
    >
      {children}
    </PopularCoursesContext.Provider>
  )
}

export const usePopularCourses = () => {
  const context = useContext(PopularCoursesContext)

  if (context === null) {
    throw new Error("PopularCoursesContext error")
  }

  return context
}
