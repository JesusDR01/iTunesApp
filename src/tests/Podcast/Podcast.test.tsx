import { render } from '@testing-library/react';

import PodcastPage from 'pages/podcast/[handle]';

import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { SearchPodcastsProvider } from 'context/search-context';
import { PlayerContextProvider } from 'context/player-context';

const queryClient = new QueryClient({
	defaultOptions: {
		queries: {
			retry: false,
		},
	},
});

const AllTheProviders = ({ children }: { children: React.ReactElement }) => {
	return (
		<QueryClientProvider client={queryClient}>
			<PlayerContextProvider>
				<SearchPodcastsProvider>{children}</SearchPodcastsProvider>
			</PlayerContextProvider>
		</QueryClientProvider>
	);
};

const DEFAULT_PODCAST_ID = '1535809341';

// Mock IntersectionObserver
class IntersectionObserver {
	observe = jest.fn();
	disconnect = jest.fn();
	unobserve = jest.fn();
}

Object.defineProperty(window, 'IntersectionObserver', {
	writable: true,
	configurable: true,
	value: IntersectionObserver,
});

Object.defineProperty(global, 'IntersectionObserver', {
	writable: true,
	configurable: true,
	value: IntersectionObserver,
});


describe('Podcast screen', () => {
	it('should render the podcast screen without crashing', () => {
		const { container } = render(
			<PodcastPage podcastId={DEFAULT_PODCAST_ID} />,
			{ wrapper: AllTheProviders },
		);
		expect(container).toBeInTheDocument();
	});

	it('should match snapshot', () => {
		const { baseElement } = render(
			<PodcastPage podcastId={DEFAULT_PODCAST_ID} />,
			{ wrapper: AllTheProviders },
		);
		expect(baseElement).toMatchSnapshot();
	});
});
