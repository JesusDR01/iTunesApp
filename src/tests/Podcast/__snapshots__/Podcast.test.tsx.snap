// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Podcast screen should match snapshot 1`] = `
<body>
  <div>
    <div>
      <div
        class="w-full"
      >
        <div
          class="flex items-center justify-center flex-col max-w-[832px] m-auto"
        >
          <section
            class="flex flex-col pr-5  items-center  mt-[20px]"
          >
            <header
              class="flex w-[794px] items-center max-w-full  justify-between gap-5 mb-6 max-md:flex-wrap max-md:justify-center"
            >
              <div
                class="flex flex-[1_0_0] w-[78px] max-w-full items-start justify-between gap-5 mt-1"
              >
                <span
                  class="text-white  flex text-opacity-30 text-base font-semibold mt-px"
                >
                  #
                </span>
                <span
                  class="flex-[1_0_0]  text-white text-opacity-30 text-sm font-semibold"
                >
                  Title
                </span>
              </div>
              <span
                class="flex-[255px_0_0] text-white text-opacity-30 text-sm font-semibold mt-px"
              >
                Topic
              </span>
              <div
                class="flex  content-between w-[154px] max-w-full items-center justify-between gap-5"
              >
                <span
                  class=" flex-[.5_0_0] text-white text-opacity-30 text-sm font-semibold mt-px"
                >
                  Released
                </span>
                <img
                  alt="Release Image"
                  class="aspect-square flex-[1_1_1] object-cover object-center w-3.5 fill-white fill-opacity-30 overflow-hidden shrink-0"
                  loading="lazy"
                  srcset="https://cdn.builder.io/api/v1/image/assets/TEMP/c036624a-8364-4c33-9d38-a53f99ad81d0?apiKey=70259793749c4d08babc2ab577934360&width=100 100w, https://cdn.builder.io/api/v1/image/assets/TEMP/c036624a-8364-4c33-9d38-a53f99ad81d0?apiKey=70259793749c4d08babc2ab577934360&width=200 200w, https://cdn.builder.io/api/v1/image/assets/TEMP/c036624a-8364-4c33-9d38-a53f99ad81d0?apiKey=70259793749c4d08babc2ab577934360&width=400 400w, https://cdn.builder.io/api/v1/image/assets/TEMP/c036624a-8364-4c33-9d38-a53f99ad81d0?apiKey=70259793749c4d08babc2ab577934360&width=800 800w, https://cdn.builder.io/api/v1/image/assets/TEMP/c036624a-8364-4c33-9d38-a53f99ad81d0?apiKey=70259793749c4d08babc2ab577934360&width=1200 1200w, https://cdn.builder.io/api/v1/image/assets/TEMP/c036624a-8364-4c33-9d38-a53f99ad81d0?apiKey=70259793749c4d08babc2ab577934360&width=1600 1600w, https://cdn.builder.io/api/v1/image/assets/TEMP/c036624a-8364-4c33-9d38-a53f99ad81d0?apiKey=70259793749c4d08babc2ab577934360&width=2000 2000w, https://cdn.builder.io/api/v1/image/assets/TEMP/c036624a-8364-4c33-9d38-a53f99ad81d0?apiKey=70259793749c4d08babc2ab577934360&"
                />
              </div>
            </header>
          </section>
          <div>
            .
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
